### PR TITLE
Fixes for admin

### PR DESCRIPTION
**Pull request by GigiusB**
Inspired by https://github.com/PropertyGuard/django-relativedelta

I also believe that we should fall back to a varchar(27) storage for db other than Postgres

### CODE REVIEW

File: relativedeltafield/__init__.py [hunk 1]
```diff
@@ -25,9 +25,31 @@
     r'$'
 )
 
+class iso8601relativedelta(relativedelta):
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:28]: What is the purpose of this class?  It seems to do nothing but pass its arguments on to the superclass, and optionally remember the original value passed in.

File: relativedeltafield/__init__.py [hunk 2]
```diff
@@ -25,9 +25,31 @@
...
+
+
 # Parse ISO8601 timespec
-def parse_relativedelta(str):
-	m = iso8601_duration_re.match(str)
+def parse_relativedelta(value):
+	if isinstance(value, relativedelta):
+		return iso8601relativedelta(value)
+	if value is None or value == '':
+		return None
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:51]: Why are these needed? The parse function should only be called on strings. See the `to_python` method of the field class, that handles `None` values.

File: relativedeltafield/__init__.py [hunk 3]
```diff
@@ -75,6 +97,45 @@ def format_relativedelta(relativedelta):
 		return 'P{}'.format(result_big)
 
 
+class RelativeDeltaFormField(forms.CharField):
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:102]: Could you add tests for the form field? This package is relatively well-tested, and I don't use Django admin myself, so without tests I'm definitely afraid to break things if I change anything.

File: relativedeltafield/__init__.py [hunk 4]
```diff
@@ -87,13 +148,15 @@ class RelativeDeltaField(models.Field):
 					 "ISO8601 interval format.")
 	}
 	description = _("RelativeDelta")
-
+	form_class = RelativeDeltaFormField
+	descriptor_class = RelativeDeltaDescriptor
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:154]: What's the descriptor for?

File: relativedeltafield/__init__.py [hunk 5]
```diff
@@ -87,13 +148,15 @@ class RelativeDeltaField(models.Field):
...
 
 	def db_type(self, connection):
-		if connection.vendor == 'postgresql':
+		if connection.vendor == 'postgresql3':
```
Comment (sjamaan) [inline relativedeltafield/__init__.py]: Why is this needed, and won't this break stuff on my own installations where this worked?

File: relativedeltafield/__init__.py [hunk 6]
```diff
@@ -87,13 +148,15 @@ class RelativeDeltaField(models.Field):
...
 			return 'interval'
 		else:
-			raise ValueError(_('RelativeDeltaField only supports PostgreSQL for storage'))
+			return 'varchar(27)'
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:160]: Let's restore the original code.  I have no intention of supporting databases other than Postgresql. This project was specifically intended to support `interval` fields on Postgres. There are no tests for other databases, and I would hate it if the above change would silently cause my fields to be silently converted to varchars.

File: relativedeltafield/__init__.py [hunk 7]
```diff
@@ -134,18 +197,21 @@ def get_db_prep_value(self, value, connection, prepared=False):
 	# We can't simply replace or remove PsycoPg2's parser, because
 	# that would mess with any existing Django DurationFields, since
 	# Django assumes PsycoPg2 returns pre-parsed datetime.timedeltas.
+	# 	def select_format(self, compiler, sql, params):
+	# 		fmt = 'to_char(%s, \'PYYYY"Y"MM"M"DD"DT"HH24"H"MI"M"SS.US"S"\')' % sql
+	# 		return fmt, params
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:203]: Please don't comment out code that's unused. Just drop it instead. It can always be retrieved from version control history.

File: relativedeltafield/__init__.py [hunk 8]
```diff
@@ -134,18 +197,21 @@ def get_db_prep_value(self, value, connection, prepared=False):
...
+
 	def select_format(self, compiler, sql, params):
-		fmt = 'to_char(%s, \'PYYYY"Y"MM"M"DD"DT"HH24"H"MI"M"SS.US"S"\')' % sql
+		if compiler.connection.vendor == 'postgresql3':
+			fmt = 'to_char(%s, \'PYYYY"Y"MM"M"DD"DT"HH24"H"MI"M"SS.US"S"\')' % sql
+		else:
+			fmt = sql
 		return fmt, params
 
-	if django.VERSION < (2,):
```
Comment (sjamaan) [inline relativedeltafield/__init__.py:141]: I'm okay with dropping older Django support (older versions are untested anyway), but let's do it in a separate commit. This is completely unrelated to the django admin support.

**Review by sjamaan — CHANGES_REQUESTED** [review]
I'm fine with adding support for Django admin, but this really needs tests. I'm not using Django-admin myself. Please note that PR #7 is for the same, but the author never got back to me. I will be unable to support Django-admin, and if there are no tests this will only make it extra hard as well.

I'd also appreciate if you could improve the commit messages a bit and perhaps rebase to make them more logical when reviewing the history.

Regarding the support for other databases, let's not do that. I see this field more as something that would belong to `django.contrib.postgres`.

File: relativedeltafield/__init__.py (hunk 1 above)
Comment (GigiusB) [reply relativedeltafield/__init__.py:28]: I thought the str() would be better to return an is88601 string format rather than "relativedelta(...)"

File: relativedeltafield/__init__.py (hunk 3 above)
Comment (GigiusB) [reply relativedeltafield/__init__.py:102]: Will do

File: relativedeltafield/__init__.py (hunk 7 above)
Comment (GigiusB) [reply relativedeltafield/__init__.py:203]: Sorry it was a left orver

File: relativedeltafield/__init__.py (hunk 8 above)
Comment (GigiusB) [reply relativedeltafield/__init__.py:141]: In reality I just simplified the if: there are no differences in the two if branches if we just leave one signature with an optional (and unused) context argument

File: relativedeltafield/__init__.py (hunk 6 above)
Comment (GigiusB) [reply relativedeltafield/__init__.py:160]: This was a proposal: we could initially fall back to a varchar with other databases and if they provide a field type similar to the Postgres interval we can think of implementing it later?

File: relativedeltafield/__init__.py (hunk 5 above)
Comment (GigiusB) [reply relativedeltafield/__init__.py]: This was a mistake fixed in a subsequent commit.

File: relativedeltafield/__init__.py (hunk 8 above)
Comment (sjamaan) [reply relativedeltafield/__init__.py:141]: hm, I don't remember.. maybe this gave a deprecation warning before. But you're right.

**Comment by GigiusB** [comment]
Closing this pull request and will push new ones following recommendatins
